File: src/numen/examples/oscillator/dynamics.py
from typing import ClassVar, Literal

import numpy as np

from numen.compiler.flatten import CompiledSpec, CompiledSystem  # noqa: F401
from numen.spec.system import System, DynamicsFn
from components import OscillatorComponent


def oscillator_dynamics(
    dx: np.ndarray,
    x: np.ndarray,
    p: np.ndarray,
    t: float,
    spec: CompiledSpec,
    system: CompiledSystem[tuple[str]],
) -> None:
    """Harmonic oscillator dynamics: ẋ = v, v̇ = -ω²x - 2ζω·v"""
    for (entity_id,) in system.entity_groups:
        c  = spec.view(entity_id, OscillatorComponent, x, p)
        dc = spec.dx_view(entity_id, OscillatorComponent, dx)
        dc.position += c.velocity
        dc.velocity += -c.omega**2 * c.position - 2 * c.damping * c.omega * c.velocity


class OscillatorSystem(System):
    component_types: ClassVar[tuple[type, ...]] = (OscillatorComponent,)
    python_fn:       ClassVar[DynamicsFn]       = staticmethod(oscillator_dynamics)
    kind:            Literal["oscillator_system"] = "oscillator_system"
    dynamics_fn:     str = "OscillatorDynamics.oscillator_dynamics!"

File: src/numen/examples/oscillator/dynamics.jl
module OscillatorDynamics

import Main: CompiledSpec, CompiledSystemSpec, groups,
             get_state, get_param, add_deriv!

"""
    oscillator_dynamics!(dx, x, p, t, spec, sys)

Harmonic oscillator: ẋ = v,  v̇ = -ω²x - 2ζωv.

Mirrors the Python ``oscillator_dynamics`` function.  Signature is identical
across Python and Julia: ``(dx, x, p, t, spec, sys)``.
"""
function oscillator_dynamics!(
    dx  :: AbstractVector{T},
    x   :: AbstractVector{S},
    p   :: Vector{Float64},
    t   :: Real,
    spec:: CompiledSpec,
    sys :: CompiledSystemSpec,
) where {T <: Real, S <: Real}
    for (eid,) in groups(sys)
        pos     = get_state(spec, x, eid, "oscillator.position")
        vel     = get_state(spec, x, eid, "oscillator.velocity")
        omega   = get_param(spec, p, eid, "oscillator.omega")
        damping = get_param(spec, p, eid, "oscillator.damping")

        add_deriv!(spec, dx, eid, "oscillator.position", vel)
        add_deriv!(spec, dx, eid, "oscillator.velocity",
                   -omega^2 * pos - 2 * damping * omega * vel)
    end
end

end  # module OscillatorDynamics
